Semperis directory services protector

ericlanteigne-semperis opened the pull request:
## Proposed Changes

  - Adding Semperis Directory Services Protector connector

ericlanteigne-semperis commented:
We are using the Event table populated by an agent collecting windows event logs. It seems there is no schema available in the tests for the Event table. The DataConnectorValidations are not listed as required but I would be happy to update the tests accordingly. If you have advice on how to get the schema for the Event table that would be very useful.

v-admahe commented on Solutions/SemperisDirectoryServicesProtector/Workbooks/Images/Logo/Semperis.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 73.6 98.2"><defs><linearGradient id="a" y1="27.34" x2="73.5" y2="27.34" gradientTransform="matrix(1, 0, 0, -1, 0, 99.84)" gradientUnits="userSpaceOnUse"><stop offset="0" stop-color="#ee4036"/><stop offset="0.19" stop-color="#ee3c3c"/><stop offset="0.43" stop-color="#ed2f4c"/><stop offset="0.71" stop-co ...
```
Id in logo file should be of guid type and xmlns:xlink should not be there in the file.

Please follow the guideline for logo [here](https://github.com/Azure/Azure-Sentinel/blob/e1665d4c4442b430d6847adc65f4b9c25f7b071f/DataConnectors/ReadMe.md)

Thanks,
Aditi Maheshwari

v-admahe replied:
@ericlanteigne-semperis Could ypu please add the sample data in the PR in json format.

Thanks,
Aditi Maheshwari

v-admahe replied:
@ericlanteigne-semperis Please move the workbook logo file in and preview images in solution folder you can take reference of folder structure from [here](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/PaloAltoPrismaCloud/Workbooks/Images). and also if its a solution workbook metadata is not needed, so could you please revert the changes of workbook metadata.

Thanks,
Aditi Maheshwari

ericlanteigne-semperis replied:
I have done these changes. Please let me know if I have missed anything.

Thank you

oshezaf commented:
Few comments:

- The parser uses mv-expand breaking each event to multiple events with a single name value pair from the XML. It is not a valid parser, even if it may work for the simplistic detections currently available. Also, parse_xml is highly inefficient. The parser should extract the relevant fields needed, preferably using parse.
- The detections use frequency of 1h and period of 5h. This will result in duplication.  the frequency and period should be the same.
- Since all detections are identical, essentially elevating an event to an alert, they should use a single rule and the new dynamic alert capabitlites.
- Is the summarization in detections intentional? Trying to solve the parser issues? If the intent is to raise an alert for each event, as I think it should be, Event grouping should be used.
- The connector lists WindowsEvent as data type while it should be Event.

v-admahe commented:
@ericlanteigne-semperis Please move the workbook logo file in and preview images in solution folder you can take reference of folder structure from here. and also if its a solution workbook metadata is not needed, so could you please revert the changes of workbook metadata.

Thanks,
Aditi Maheshwari

v-admahe commented:
@ericlanteigne-semperis Could ypu please add the sample data in the PR in json format.

Thanks,
Aditi Maheshwari

ericlanteigne-semperis commented:
I will reopen this pull request when we are done recommended changes

ericlanteigne-semperis commented on Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_EvidenceOfMimikatzDCShadowAttack.yaml:
```diff
@@ -0,0 +1,23 @@
+requiredDataConnectors:
+  - connectorId: SemperisDSP
+    dataTypes:
+      - WindowsEvent
```
Can we resolve this comment?

ericlanteigne-semperis commented on Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_KerberoskrbtgtAccount.yaml:
```diff
@@ -0,0 +1,23 @@
+requiredDataConnectors:
+  - connectorId: SemperisDSP
+    dataTypes:
+      - WindowsEvent
```
Can we resolve this comment?

ericlanteigne-semperis commented on Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_RecentsIDHistoryChangesOnADObjects.yaml:
```diff
@@ -0,0 +1,23 @@
+requiredDataConnectors:
+  - connectorId: SemperisDSP
+    dataTypes:
+      - WindowsEvent
```
Can we resolve this comment?

ericlanteigne-semperis commented on Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_WellKnownPrivilegedSIDsInsIDHistory.yaml:
```diff
@@ -0,0 +1,24 @@
+requiredDataConnectors:
+  - connectorId: SemperisDSP
+    dataTypes:
+      - WindowsEvent
```
Can we resolve this comment?

ericlanteigne-semperis commented on Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_ZerologonVulnerability.yaml:
Can we resolve this comment?

ericlanteigne-semperis commented on Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json:
```diff
@@ -0,0 +1,152 @@
+    "title": "Semperis Directory Services Protector",
+    "publisher": "SEMPERIS",
+    "descriptionMarkdown": "Semperis Directory Services Protector data connector allows for the export of its Windows event logs (i.e. Indicators of Exposure and Indicators of Compromise) to Azure Sentinel in real time.\nIt provides a data parser to manipulate the Windows event logs more easily. The different workbooks ease your Active Directory seurity monitoring and provide different ways to vi ...
+    "additionalRequirementBanner": "This data connector depends on a parser based on Kusto Function to work as expected. Follow the steps to use this Kusto Function alias **dsp_parser** in queries and workbooks. [Follow steps to get this Kusto Function>](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/SemperisDirectorySerivesProtector/Parsers/dsp_parser.txt)  ",
```
Can we resolve this comment?

ericlanteigne-semperis commented on Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json:
```diff
@@ -0,0 +1,152 @@
+    "instructionSteps": [
+        {
+            "title": "", 
+            "description": ">This data connector depends on a parser based on a Kusto Function to work as expected. [Follow these steps](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/SemperisDirectorySerivesProtector/Parsers/dsp_parser.txt) to create the Kusto Functions alias, **dsp_parser**",
```
Can we resolve this comment?

ericlanteigne-semperis commented:
@v-admahe I've added the Sample Data. I can see that most of the other sample data files are CSV. I added that one as well since I wasn't sure. In the ReadMe for Solutions it says to put it in the appropriate folder but seeing as our parser is using the Event table, I am not certain where it should go. Do you have advice?
Also I removed the changes from Workbooks directory and kept everything in the Solutions directory.

ericlanteigne-semperis commented:
@oshezaf Do you have more information on "the new dynamic alert capabilities"? I wasn't quite able to find information on that just yet. That would be helpful, thanks. 
At the moment we just want a single alert to be triggered when the detection query has results, so I've removed the summarize statement. Based on your message it sounds like there is a way to trigger multiple alerts when running a single detection, is there more information about this in documentation? I am not seeing information stating that a single rule can have multiple alerts but I am curious about it. 
I am reviewing how the parser is working to see if I can use parse and avoid mx-expand however the information in EventData does not have a defined schema so I am not certain we will be able to make that change to parse_xml. Is there an issue if we have to use parse_xml? 
Also, we are using pivot to recombine the results that were expanded via mx-expand based on the values that were common before the expansion. Is this not an adequate method of maintaining data consistency?

oshezaf commented:
@ericlanteigne-semperis :

- Event grouping (i.e. generating an alert per query result): Just added documentation on that here: https://github.com/Azure/Azure-Sentinel/wiki/Query-Style-Guide#eventgroupingsettings.

- Pivot will not actually regenerate the events. It will group them based on the dimensions you specify, but since there is no unique key dimension, it is not deterministic. In practice you pivot on time with the risk of two events at the same time being grouped together. It is also highly inefficient for parsers and will break for high rates of events. I am aware there are other parsers using this method, but it is flawed. As to the alternative: parse is preferable for any part of the XML that has repeating structure. You can have multiple parse statements and use extract/extract_all for variable parts. Can you share XML samples?

ericlanteigne-semperis commented:
@oshezaf Thank you for the information. 

I included sample data in both csv and json format. I've simply put it in the Sample Data directory for now. @v-admahe asked for it but I still need to confirm where exactly I should put it

ericlanteigne-semperis commented:
I have an issue open with regards to the WorkbooksValidations failing.
https://github.com/Azure/Azure-Sentinel/issues/2389

Also, the sample data I added is causing the DocumentsLinkValidation to fail. Any advice on where that should be located? Inside the Solution folder potentially?

Lastly I haven't gotten any feedback regarding the DataConnectorValidations that are failing. I included a comment originally with regards to this, if you have any advice on this as well that would be much appreciated

> We are using the Event table populated by an agent collecting windows event logs. It seems there is no schema available in the tests for the Event table. The DataConnectorValidations are not listed as required but I would be happy to update the tests accordingly. If you have advice on how to get the schema for the Event table that would be very useful.

Thank you

ericlanteigne-semperis commented on Solutions/SemperisDirectoryServicesProtector/Parsers/dsp_parser.txt:
```diff
@@ -0,0 +1,30 @@
+| extend EventData = parse_xml(EventData).DataItem.EventData.Data
+| mv-expand bagexpansion=array EventData
+| evaluate bag_unpack(EventData)
+| extend Key=tostring(['@Name']), Value=['#text']
```
The parser has been updated and this is no longer used. Can this comment be resolved?

Thanks

ericlanteigne-semperis commented:
Going to open a new clean pull request since this one seems to have gotten too old and nobody is replying to my comments
